openfeature/multiprovider/multiprovider.go

```diff
@@ -0,0 +1,567 @@
+	MetadataSuccessfulProviderName         = "multiprovider-successful-provider-name"
+	MetadataSuccessfulProviderNames        = MetadataSuccessfulProviderName + "s"
+	MetadataStrategyUsed                   = "multiprovider-strategy-used"
+	MetadataFallbackUsed                   = "multiprovider-fallback-used"
+	MetadataIsDefaultValue                 = "multiprovider-is-result-default-value"
+	MetadataEvaluationError                = "multiprovider-evaluation-error"
+	MetadataComparisonDisagreeingProviders = "multiprovider-comparison-disagreeing-providers"
+)
+
+type (
+	// ProviderMap is an alias for a map containing unique names for each included [of.FeatureProvider]
+	ProviderMap = map[string]of.FeatureProvider
+
+	// MultiProvider is an implementation of [of.FeatureProvider] that can execute multiple providers using various
+	// strategies.
+	MultiProvider struct {
+		providers          ProviderMap
+		metadata           of.Metadata
+		initialized        bool
+		totalStatus        of.State
+		totalStatusLock    sync.RWMutex
+		providerStatus     map[string]of.State
+		providerStatusLock sync.Mutex
+		strategyName       EvaluationStrategy    // the name of the strategy used for evaluation
+		strategy           StrategyFn[FlagTypes] // used for custom strategies
+		logger             *slog.Logger
+		outboundEvents     chan of.Event
+		inboundEvents      chan namedEvent
+		workerGroup        sync.WaitGroup
+		shutdownFunc       context.CancelFunc
+		globalHooks        []of.Hook
+	}
+
+	// NamedProvider allows for a unique name to be assigned to a provider during a multi-provider set up.
+	// The name will be used when reporting errors & results to specify the provider associated.
+	NamedProvider struct {
+		Name string
+		of.FeatureProvider
+	}
+
+	// configuration is the internal configuration of a [MultiProvider]
+	configuration struct {
+		useFallback      bool
+		fallbackProvider of.FeatureProvider
+		customStrategy   StrategyFn[FlagTypes]
+		logger           *slog.Logger
+		timeout          time.Duration
+		hooks            []of.Hook
+		providerHooks    map[string][]of.Hook
+		customComparator Comparator
+	}
+
+	// Option function used for setting configuration via the options pattern
+	Option func(*configuration)
+
+	// Private Types
+	namedEvent struct {
+		of.Event
+		providerName string
+	}
+)
+
+var (
+	_                of.FeatureProvider = (*MultiProvider)(nil)
+	_                of.EventHandler    = (*MultiProvider)(nil)
+	_                of.StateHandler    = (*MultiProvider)(nil)
+	stateValues      map[of.State]int
+	stateTable       [3]of.State
+	eventTypeToState map[of.EventType]of.State
+)
+
+// init Initialize "constants" used for event handling priorities and filtering.
+func init() {
+	// used for mapping provider event types & provider states to comparable values for evaluation
+	stateValues = map[of.State]int{
+		"":            -1, // Not a real state, but used for handling provider config changes
```

Comment:
[curiosity] could we name this? - most likely not, but it would be cool to have some kind of named string here, for clarity